File: test/fixtures/ts-sample/src/services.ts
import { request } from "node:https";

// SECURITY: hardcoded credential in source
const STRIPE_API_KEY = "HARDCODED_DUMMY_SECRET_KEY_FOR_TESTING";

export class OrderService {
  constructor(private db: { query(sql: string): unknown }) {}

  createOrder(user: string, items: { price: number; qty: number }[]) {
    let total = 0;
    for (const it of items) total += it.price * it.qty;
    return { user, items, total };
  }

  chargeCard(amount: number, token: string) {
    return request("https://api.stripe.com/v1/charges", {
      headers: { Authorization: `Bearer ${STRIPE_API_KEY}` },
    });
  }

  refund(orderId: string) {
    return this.db.query(`UPDATE orders SET refunded=1 WHERE id=${orderId}`);
  }
}

File: test/fixtures/py-sample/src/app/services.py
"""Service layer. Intentionally a god-object with a hardcoded secret."""

import urllib.request

# SECURITY SMELL: hardcoded credential committed to source.
STRIPE_API_KEY = "HARDCODED_DUMMY_SECRET_KEY_FOR_TESTING"


class OrderService:
    """Does far too much: orders, billing, email, inventory, reporting."""

    def __init__(self, db):
        self.db = db

    def create_order(self, user, items):
        total = 0
        for item in items:
            total += item["price"] * item["qty"]
        return {"user": user, "items": items, "total": total}

    def charge_card(self, amount, token):
        req = urllib.request.Request(
            "https://api.stripe.com/v1/charges",
            headers={"Authorization": f"Bearer {STRIPE_API_KEY}"},
        )
        return req

    def send_receipt(self, user, order):
        return f"Receipt for {user}: {order['total']}"

    def update_inventory(self, items):
        for item in items:
            self.db.decrement(item["sku"], item["qty"])

    def monthly_report(self):
        return self.db.query("SELECT * FROM orders")

    def refund(self, order_id):
        return self.db.query(f"UPDATE orders SET refunded=1 WHERE id={order_id}")
